src/Uno.UI.RemoteControl.VS/Commands/UnoMenuCommand.cs

```diff
@@ -0,0 +1,126 @@
+
+		if (Instance is null
+			&& await package.GetServiceAsync(typeof(IMenuCommandService)) is OleMenuCommandService commandService)
+		{
+			Instance = new UnoMenuCommand(package, ideChannelClient, commandService, cr);
+
+			CommandID unoMainMenuId = new CommandID(UnoStudioPackageCmdSet, UnoMainMenu);
+			if (Instance.CommandService.FindCommand(unoMainMenuId) is not OleMenuCommand)
+			{
+				var unoMenuItem = new OleMenuCommand(null, unoMainMenuId);
+				unoMenuItem.BeforeQueryStatus += Instance.OnBeforeQueryStatus;
+				commandService.AddCommand(unoMenuItem);
+			}
+
+			CommandID dynamicMenuCommandIdId = new CommandID(UnoStudioPackageCmdSet, DynamicMenuCommandId);
+			if (Instance.CommandService.FindCommand(dynamicMenuCommandIdId) is DynamicItemMenuCommand dynamicMenuItem)
+			{
+				dynamicMenuItem.BeforeQueryStatus += Instance.OnBeforeQueryStatus;
+			}
+		}
+	}
+
+	private void OnBeforeQueryStatus(object sender, EventArgs e)
+	{
+		var command = sender as OleMenuCommand;
+		if (command != null && command.Visible == false)
+		{
+			command.Visible = true;
+			command.Enabled = true;
+		}
+	}
+
+	/// <summary>
+	/// True if the command ID is a valid dynamic item
+	/// [inside the range of itens, from DynamicMenuCommandId to CommandList.Count]
+	/// </summary>
+	private bool IsValidDynamicItem(int commandId)
+		=> (commandId >= (int)DynamicMenuCommandId) &&
+			((commandId - (int)DynamicMenuCommandId) < CommandList.Count);
+
+	private void OnInvokedDynamicItem(object sender, EventArgs args)
+	{
+		if (IdeChannelClient != null &&
+				sender is DynamicItemMenuCommand matchedCommand &&
+				TryGetCommandRequestIdeMessage(matchedCommand, out var currentCommand))
+		{
+			// Ignoring the async call as the OleMenuCommand.execHandler is private and not async
+			_ = IdeChannelClient.SendToDevServerAsync(currentCommand, _package.DisposalToken);
+		}
+	}
+
+	private void OnBeforeQueryStatusDynamicItem(object sender, EventArgs args)
+	{
+		ThreadHelper.ThrowIfNotOnUIThread();
+		if (!CommandList.Any())
+		{
+			return;
+		}
+		DynamicItemMenuCommand matchedCommand = (DynamicItemMenuCommand)sender;
+		matchedCommand.Enabled = true;
+		matchedCommand.Visible = true;
+
+		if (TryGetCommandRequestIdeMessage(matchedCommand, out var currentCommand))
+		{
+			matchedCommand.Text = currentCommand.Command;
+		}
+		// Clear the ID because we are done with this item.
+		matchedCommand.MatchedCommandId = 0;
+	}
+
+	private static int GetCurrentPosition(DynamicItemMenuCommand matchedCommand) =>
+			// The position of the command is the command ID minus the ID of the root dynamic start item.
+			matchedCommand.MatchedCommandId == 0 ? 0 : matchedCommand.MatchedCommandId - (int)DynamicMenuCommandId;
+
+	private bool TryGetCommandRequestIdeMessage(DynamicItemMenuCommand matchedCommand, [NotNullWhen(true)] out CommandRequestIdeMessage result)
+		=> (result = CommandList.Skip(GetCurrentPosition(matchedCommand)).FirstOrDefault()) != null;
```

Comment:
We are matching command by position in the queue? We are not able to have a "state" object in the menu items pushed in VS ?

Reply:
We are using the Guid from the Reference Menu to grab the elements and update the name and respective command.

Using the Menu and the Next X menus, considering the size of the List of items (CommandList)

Following the example in the link [Dynamically add menu items](https://learn.microsoft.com/en-us/visualstudio/extensibility/dynamically-adding-menu-items?view=vs-2022#add-the-command)

but I can think of something different if you wish.